### PR TITLE
Add factorial circuit

### PR DESCRIPTION
**Pull request by sraver**
Adds an example of a circuit implementation that computes the factorial any given number, to a certain maximum.

Using Python frontend to verify the approach, will continue to port the example to Rust.

### CODE REVIEW

**Review by leolara — APPROVED** [review]

**Comment by leolara** [comment]
@sraver approved, please merge

**Comment by sraver** [comment]
@leolara included the same example in rust too

File: examples/factorial.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    // ...
+
+    use chiquito::frontend::dsl::cb::*; // functions for constraint building
+
+    let factorial_circuit = circuit::<F, u32, _>("factorial", |ctx| {
+        let i = ctx.shared("i");
+        let x = ctx.forward("x");
+
+        // setup step will make sure the circuit is initialized correctly
+        let setup_step = ctx.step_type_def("setup_step", |ctx| {
```
Comment (leolara) [inline examples/factorial.rs:47]: The use of the word setup can be confusing

Comment (sraver) [reply examples/factorial.rs]: I see the problem. Updated to `FirstStep` now :+1:

**Pull request merged**